Disable escaping from namespaces for inner functions (#208)

ChengCat opened the pull request:
The namespace escaping behaviour has been removed, and a new machanism for checking invalid variable reference is added.

This is mostly done, and the error reporting could be improved:
* Error location could be added.
* I don't understand for now why `error-src/anonymous-scopes.dt` generates two "can't reference variable in a different function" errors.

I tackle this issue (#208) first instead of #201, because this seems to be easier, and it's more urgent for my use.

tomhrr commented:
Thanks for this, it looks good.  The only change I'd like to make is to drop the `ReferenceVariableInDifferentFunction` error, because it works differently from the existing 'variable not in scope' errors, and those existing errors will be displayed whenever the new errors are displayed anyway.  Does that sound OK?

ChengCat commented:
Thanks for the feedback. I originally added that error, because my mental model is that the variable is in the scope, but can't be referenced as a value. (Function scope is viewed as a compiler implementation detail.)
I have removed the error, and changed `VariableNotInScope`'s error message to indicate the different possible cause. Does it look OK?

tomhrr commented:
Yep, these changes look good, thanks.
